src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/Workspace/Mef/MefWorkspaceServices.cs

```diff
@@ -51,19 +57,76 @@ public override HostServices HostServices
 
         public override Workspace Workspace => _workspace;
 
+        public override void Dispose()
+        {
+            ImmutableArray<IDisposable> disposableServices;
+            lock (_gate)
+            {
+                disposableServices = _ownedDisposableServices.ToImmutableArray();
+                _ownedDisposableServices.Clear();
+            }
+
+            // Take care to give all disposal parts a chance to dispose even if some parts throw exceptions.
+            List<Exception> exceptions = null;
+            foreach (var service in disposableServices)
+            {
+                try
+                {
+                    service.Dispose();
+                }
+                catch (Exception ex) when (FatalError.ReportWithoutCrashUnlessCanceledAndPropagate(ex))
```

Comment:
if a service cancels during dispose... is that not busted?

Reply:
This is just telemetry. I can change this to include cancellation exceptions in telemetry.